Telefon doğrulama servisi ekle ve yerli sms sağlayıcıları kullan

ecylmz opened the pull request:
#### Bu PR'in yaptığı işi/değişikliği ve bu işi/değişikliği neden yaptığını açıklayın

Twilio'nun sms ücretlerinin yerli sms firmalarına göre çok pahalı olmasından dolayı, doğrulama işlemlerinde gönderilen sms'ler yerli firmalar üzerinden gönderilmeli. Bu aşamada @roktas'ın yazmış olduğu sms gem'i kullanıldı. https://github.com/omu/sms

Bu gem aracılığıyla tek bir sms sağlayıcısına bağımlı olmaktan kurtuluyoruz. Şu anda İleti Merkezi, MutluCell, Vatan SMS ve Verimor sağlayıcılarını destekliyor. Bu çalışmada ileti merkezi kullanıldı.

Ayrıca doğrulama aşamasında Twilio Verify kullanıldığı için bu servisin benzeri Nokul içerisinde gerçeklendi. Telefon doğrulamaları artık `VerificationService` üzerinden yapılabilir.

Doğrulama aşamasında kodlar ve limit işlemleri için Redis kullanıldı.

#### İlgili/kapatılacak iş kayıtları

#### Teknik borç kayıtları

#### Veritabanına etkileri

#### Sistem etkileri

#### Kontrol listesi

- [x] [Katkı sağlama dokümanını](../blob/master/.github/CONTRIBUTING.md) okudum
- [x] İş kaydının başlığı kurallara (sadece ilk harf büyük, emir kipinde problem cümlesi) uygun
- [x] Yapılan iş/değişikliği dokümante ettim
- [x] Yapılan iş/değişikliğin testlerini yazdım
- [x] Test coverage oranını kontrol ettim
- [x] Kod kalitesi (karma) ve test suite dahil olmak üzere tüm entegre kontroller başarıyla geçiyor
- [x] Kendimi bu PR'e assign ettim
- [x] Yapılan iş/değişiklik ile ilgili proje üyelerinden review talep ettim
- [x] Gerekli etiketlemeyi (ör. bug) yaptım

#### Ek içerik

[//]: # (Kaynaklar)
[//]: # (Dış bağlantılar)
[//]: # (Ekran görüntüleri)
[//]: # (Örnek çözümler)

ecylmz committed:
"SMS gem'ini ve sağlayıcı kimlik bilgilerini ekle"

ecylmz committed:
Verification modeli ekle

ecylmz committed:
Telefon doğrulama servisi ekle

ecylmz committed:
Twilio yerine Verification servisini kullan

ecylmz committed:
Verification modelini sil

ecylmz committed:
s/verify/verification

ecylmz committed:
Doğrulama servisinde redis kullan

ecylmz committed:
Remove verifications fixtures

ecylmz committed:
Update doc

roktas reviewed (commented):
@ecylmz Eline sağlık!  PR'ı biraz bekletip zaman bulduğumda bakacağım.

roktas committed:
Parola sıfırlama kodunu refaktörle (test ve review gerekiyor)

Emre'nin (@ecylmz) onayıyla İrfan (@isubas) ile istişare ederek yapılan
bu refaktörde temel amaç "Actions" kurgusunu kullanmak.  Kod test
edilmedi, ilk denemede çalışmayacağına veya CI'ın hata vereceğine
neredeyse eminim.  Review'de tartışalım.

roktas commented:
@ecylmz sen testlerini tamamlayıncaya kadar PR'ı bir süre draft'a çekelim.

roktas committed:
Hata düzelt (varan 1)

roktas committed:
Artık gerekmeyen servisi kaldır

roktas committed:
Hataları düzelt (varan 2)

roktas committed:
Çevirileri düzenle

roktas committed:
Hata düzelt

roktas committed:
Actions PORO'ları toplayan bir konvansiyon uygula

Beklediğimiz düzen şöyle (actions dizininin app/services altından app
altına taşınacağını varsayıyoruz):

	app/actions/«ilişkili model yolu»/«eylem grup adı»/«eylemler»...

Örneğin

	app/actions/prospective_student/upsert.rb

dosyası aday öğrencinin güncellenmesi (insert/update) eylemine karşı
düşüyor.

Bu örnekte ise (yer ve isimler tartışılabilir)

	app/actions/user/verification/send.rb

dosyası kullanıcının (model yolu) doğrulanması (eylem grup adı)
eylemlerinden "doğrulama kodunun gönderilmesi" eylemine karşı düşüyor.

Bu düzenlemede tecrübe ettiğimiz bir sıkıntı eylemlerde (send.rb,
verify.rb) kullanılacak PORO ve yardımcıları nereye koyacağımız
konusuydu.  Başlangıçta yaptığım başarısız bir girişimden sonra şöyle
bir konvansiyonun uygun olacağını düşündüm.

	app/actions/«ilişkili model yolu»/«eylem grup adı».rb

Buna göre

	app/actions/user/verification.rb

dosyası ilgili eylem grubundaki tüm PORO ve helperları barındırıyor ve

	app/actions/user/verification/

dizini sadece eylemlerden oluşuyor (keşfedilebilirliği artıran bir
husus).

Bu düzenlemenin zaafı birden fazla PORO class ve yardımcı varsa (ki bu
örnekte var) hepsinin tek dosyada konsolide edilmesi.  Fakat duruma göre
bu bir zaaf değil avantaj olarak da değerlendirilebilir: kullanılan tüm
yapıtaşlarının tek bir dosyada incelenebilmesi.

Alternatifler:

- PORO ve helperlar konsolide edilmeden eylem grubu dizininde kendi
  dosyalarına konulabilirdi.  Örneğin phone nesnesi:

	app/actions/user/verification/phone.rb

  Bu düzenlemedeki sıkıntı bu şekilde ortaya çıkan irili ufaklı
  dosyaların gerçek eylem dosyalarının görünürlüğünü azaltması.

	app/actions/user/verification/phone.rb # bir PORO
	app/actions/user/verification/send.rb  # bir eylem

- PORO ve helperlar eylem grubu dizininde tek bir dosyada konsolide
  edilebilirdi.  Örneğin:

	app/actions/user/verification/support.rb # veya utils.rb

  Bu düzenlemedeki sıkıntılar ise (1) ilk alternatif kadar olmasa bile
  eylem görünürlüğünü bir parça azaltması, (2) support, utils gibi çok
  jenerik çağrışımlı isimlendirmeler kullanılması.

Son olarak...  Giderek büyüme eğilimi gösteren Rails uygulamalarında
okunurluk ve maintain edilebilirliği korumak/ölçeklemek gerçekten ciddi
bir sorun.  Amaca uygun PORO kullanımı ve buna benzer konvansiyonlar bu
mücadeledeki en büyük silahımız.  Maalesef Rails bu konuda (concerns
dışında) açık ve dokümante edilmiş konvansiyonlar sunmuyor.  Bu nedenle
bu konuda net ve herkes tarafından kullanılan bir "best practice" yok
(araştırmalarım sonucunda ulaştığım yargı), fakat esinlenilebilecek
güzel uygulamalar var.  Örneğin exercism.io'nun arkasındaki

	https://github.com/exercism/v3-website

projesi (app/commands dizini; bizdeki actions ile benzer).  Sonuç olarak
bu konu tartışmaya açık.  Farklı fikirleriniz varsa dinlemek isterim.

roktas committed:
Testlerin yerini hazırla

ecylmz committed:
Telefon numarasını kullanıcı üzerinden al.

roktas committed:
Code nesnesini tam teşekküllü bir nesneye dönüştür

Yapılan testlerde Redis'ten okuma/yazma yaparken verinin doğru
serileştirmemesinden kaynaklanan bir hata oluştu (ör. string ile bir
sayının toplanması).  Oraya buraya `to_i`, `to_s` koymaktansa Code
nesnesini serialize ve deserialize yapacak şekilde geliştirdik.

Kıssadan hisse: "primitive obsession"dan kaçarken bu işi eksiksiz
yapalım.

roktas committed:
Merge branch 'jira-65' of https://github.com/omu/nokul into jira-65

ecylmz committed:
s/TooManyRequestError/TooManySendRequestError

ecylmz committed:
Parola sıfırlama sırasında telefonu kullanıcıdan al

ecylmz committed:
Telefon doğrulama testlerini ekle

ecylmz committed:
İsim çözümleme hatalarını düzelt

roktas commented:
@ecylmz Eline sağlık!  Bunu review edip gün içinde alalım.

huseyin commented on app/services/actions/prospective_student/registration.rb:
```diff
@@ -89,7 +89,7 @@ def student # rubocop:disable Metrics/MethodLength
       def user
         @user ||= begin
-          User.find_by(id_number: id_number) || User.new(
+          ::User.find_by(id_number: id_number) || ::User.new(
```
Bunun için `find_or_initialize_by` kullanabiliriz: https://apidock.com/rails/v4.0.2/ActiveRecord/Relation/find_or_initialize_by

roktas replied:
@huseyin good catch!  Source şu:

```ruby
# File activerecord/lib/active_record/relation.rb, line 200
    def find_or_initialize_by(attributes, &block)
      find_by(attributes) || new(attributes, &block)
    end
```

isubas replied:
find_or_initialize_by burası için uygun değil. Değişmeyeceğinden emin olduğun durumlar için kullanabilirsin. Yani instance oluşturulurken farklı öznitelikler var ve find işlemide bu öz niteliklere göre arama yapacağından her zaman doğru çalışmayacaktır.

roktas replied:
find_by ve new farklı argümanlar alıyor bu kodda, birebir karşılamıyor evet

isubas replied:
Ezbere yazıyorum ama,

```ruby
user = ::User.find_or_create_by(id_number: id_number) do |record|
    record.assign_attributes(
        email:                 email,
        password:              id_number,
        password_confirmation: id_number
    )
end
```

bu şekilde çalışabilir. Emin değilim. Test etmedim.

roktas replied:
Bence find_or_initialize_by ya hep ya hiç bir metod.  Olduğu gibi yazılamayacaksa eskisi daha okunur, müdahale etmeyelim.

huseyin replied:
Block'lu hali de çalışır bu arada. Kullanımını daha önce gördüm (gitlab'in source kodundaydı sanırım)

roktas commented:
gençler, approve/reject bekliyorum...
